### PR TITLE
[Feature-#8318] [master-server] Sub process transfer parameters to father process

### PR DESCRIPTION
**Pull request by wangxj3**
[Feature-#8318] [master-server] Sub process transfer parameters to father process
When you use the out param in taskNode of subpeocess, the father processInstance could get the param definde in the subProcess.

### CODE REVIEW

**Comment by caishunfeng** [comment]
Hi @wangxj3 please resolve the conflicts.

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java [hunk 1]
```diff
@@ -471,6 +472,9 @@ private void taskFinished(TaskInstance taskInstance) {
             completeTaskMap.put(taskInstance.getTaskCode(), taskInstance.getId());
         }
 
+        if(taskInstance.isSubProcess()) {
```
Comment (caishunfeng) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java]: Is there some logic lack?

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java [hunk 2]
```diff
@@ -1279,6 +1283,24 @@ private void submitPostNode(String parentNodeCode) {
             TaskInstance task = createTaskInstance(processInstance, taskNodeObject);
             taskInstances.add(task);
         }
+        //the end node of the branch of the dag
+        if (StringUtils.isNotEmpty(parentNodeCode)&&dag.getEndNode().contains(parentNodeCode)){
```
Comment (caishunfeng) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java]: please format the code

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/task/SubTaskProcessor.java [hunk 3]
```diff
@@ -110,9 +119,35 @@ private void updateTaskState() {
         if (subProcessInstance != null && subProcessInstance.getState().typeIsFinished()) {
             taskInstance.setState(subProcessInstance.getState());
             taskInstance.setEndTime(new Date());
+            dealFinish();
             processService.saveTaskInstance(taskInstance);
         }
     }
+    private void dealFinish() {
```
Comment (caishunfeng) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/task/SubTaskProcessor.java:130]: I think it's better to add some doc for `dealFinish`, what do you think?

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/task/SubTaskProcessor.java [hunk 4]
```diff
@@ -110,9 +119,35 @@ private void updateTaskState() {
...
+        String thisTaskInstanceVarPool = taskInstance.getVarPool();
+        if (StringUtils.isNotEmpty(thisTaskInstanceVarPool)) {
+            String subProcessInstanceVarPool = subProcessInstance.getVarPool();
+            if (StringUtils.isNotEmpty(subProcessInstanceVarPool)) {
+                List<Property> varPoolProperties = JSONUtils.toList(thisTaskInstanceVarPool, Property.class);
+                Map<String, Object> taskParams = JSONUtils.parseObject(taskInstance.getTaskParams(), new TypeReference<Map<String, Object>>() {
+                });
+                Object localParams = taskParams.get(LOCAL_PARAMS);
```
Comment (caishunfeng) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/task/SubTaskProcessor.java:138]: Should global parameters be handled as well?

**Comment by caishunfeng** [comment]
Hi @wangxj3 this feature also need to change doc, so I add the `miss document` label for it.

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java (hunk 1 above)
Comment (wangxj3) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java]: I will delete it

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java (hunk 2 above)
Comment (wangxj3) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java]: OK

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/task/SubTaskProcessor.java (hunk 3 above)
Comment (wangxj3) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/task/SubTaskProcessor.java:130]: OK

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/task/SubTaskProcessor.java (hunk 4 above)
Comment (wangxj3) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/task/SubTaskProcessor.java:138]: the globalParam is a static param for the instance ,so i don't think it's should be handled here.

**Review by caishunfeng — COMMENTED** [review]
LGTM

**Review by caishunfeng — APPROVED** [review]

**Review by lenboo — APPROVED** [review]
+1

**Pull request merged**